065-topic-replication-factor-change.md

```diff
@@ -0,0 +1,289 @@
+This is far from user friendly, and the user is in charge of doing the calculations to determine the best broker to host a new replica, or which replica needs to be dropped.
+
+When data loss is not a problem, the user can simply delete and recreate the KafkaTopic resource with the desired replication factor.
+This is not really an option for most use cases.
+
+## Motivation
+
+The topic replication factor change feature will provide a much better cloud-native experience for Strimzi users.
+They will be able to change the topic replication factor by simply updating the KafkaTopic resource, which may also happen through a GitOps pipeline.
+
+This is a non exhaustive list of use cases that will become significantly easier:
+
+- Set a lower replication factor for non-critical topics to reduce the infrastructure costs (compute, storage and network).
+- Set a lower replication factor to deal with resource shortage (trade-off between fault-tolerance and resource utilization).
+- Set a higher replication factor for critical topics to improve durability and fault tolerance.
+
+A tighter integration with Cruise Control is also one of the long term goals of the Strimzi project. 
+
+## Proposal
+
+The replication factor change feature can be implemented by integrating the Topic Operator with Cruise Control.
+The Cluster Operator is already integrated with Cruise Control, so we can factor out the existing logic and classes.
+
+Cruise Control creates a task for each operation, such as topic configuration and cluster rebalance.
+A task is first created in the Active state, then moves to InExecution, and finally to Completed or CompletedWithError.
+At most `max.active.user.tasks` tasks can be Active at the same time (5 by default), and only one task can be InExecution at any given time.
+During the execution, a task is divided in one or more batches, which are then executed one after the other.
+A task may include multiple topics and it is marked as completed once the whole task's status is Completed.
+
+### Topic configuration
+
+The [`topic_configuration`](https://github.com/linkedin/cruise-control/wiki/REST-APIs#change-kafka-topic-configuration) endpoint is used to request a topic configuration update, and currently only supports replication factor changes.
+Once accepted, the replication factor change is put into the Cruise Control's task queue and executed asynchronously.
+
+The two required parameters for this endpoint are `topic` (regular expression to specify subject topics) and `replication_factor` (target replication factor).
+In order to group multiple changes in the same request to Cruise Control, the Topic Operator uses a [JSON payload](https://github.com/linkedin/cruise-control/wiki/Change-topic-replication-factor-through-Cruise-Control#instruction) containing topics grouped by their replication factor, which replaces the required parameters.
```

Comment:
"containing topics ... this could be misleading. It could mean you need topic names there, you can specify a regex instead (as for the `topic` parameter). Let's make it clearer.

Reply:
Right. Updated. Let me know if it's clear enough.

Reply:
I don't know the internal details but are we sure that the payload can contain just regex like `topic1|topic2|topic3` or even more generic regex.

Reply:
Yes, it can contain any valid regex, but I'm illustrating our usage.

Reply:
Sorry I read it as an overall description of the API and not referring to our usage. Maybe we can make it clearer?

Reply:
but it says "the Topic Operator uses..."

Reply:
Right. That's fine then.